docs: tw691-arbitrum-chain-restructure (#3341)
OffchainLabs/arbitrum-docs: docs/launch-arbitrum-chain/chain-config/core/arbos.mdx

content_type: how-to
---

import CustomizeCautionBannerPartial from '../../partials/_arbitrum-chain-customize-caution-banner-partial.mdx';

<CustomizeCautionBannerPartial />

> Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer) that activates custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.

Reviewer: ```suggestion
Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer). This activates custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.
```

> Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer) that activates custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.

Reviewer: ```suggestion
Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer). This lets you activate custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.
```

> Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer) that activates custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.

Reviewer: ```suggestion
Customizing the [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx) version involves modifying the Nitro codebase to introduce a new, version-controlled iteration of ArbOS (<a data-quicklook-from="arbitrum">Arbitrum</a>'s operating system-like layer). This new iteration activates custom behaviors, features, or state changes while maintaining backward compatibility and determinism for fraud proofs.
```

> This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31) to incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables, without conflicting with official releases.

Reviewer: ```suggestion
This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31). These versions incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables, without conflicting with official releases.
```

> This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31) to incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables, without conflicting with official releases.

Reviewer: ```suggestion
This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31). This lets you incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables without conflicting with official releases.
```

> This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31) to incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables, without conflicting with official releases.

Reviewer: ```suggestion
This is an advanced customization primarily for live chains, allowing developers to extend or alter the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF)—the core logic for block production and state updates—while ensuring safe, non-disruptive upgrades. Unlike standard upgrades (e.g., to canonical versions like ArbOS 20 "Atlas"), customization involves creating intermediate or project-specific versions (e.g., ArbOS 32, a fork of 31) to incorporate bespoke elements such as new precompiles, EVM opcodes, or state variables without conflicting with official releases.
```

> It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.

Reviewer: ```suggestion
It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain reorgs or failed fraud proofs.
```

> It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.

Reviewer: ```suggestion
This approach is recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.
```

> It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.

Reviewer: ```suggestion
This customization is recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.
```

> It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain re-orgs or failed fraud proofs.

Reviewer: ```suggestion
It's recommended for teams with expertise or partners (e.g., Rollup-as-a-Service providers), as it requires audits, maintenance, and careful handling to avoid issues like chain reorgs or failed fraud proofs.
```

> ### 1. Add a new method to existing precompile on a specific ArbOS version:

Reviewer: ```suggestion
### 1. Add a new method to existing precompile on a specific ArbOS version
```

> ### 1. Add a new method to existing precompile on a specific ArbOS version:

Reviewer: ```suggestion
### 1. Add a new method to existing precompile on a specific ArbOS version
```

> ### 1. Add a new method to existing precompile on a specific ArbOS version:

Reviewer: ```suggestion
### 1. Add a new method to existing precompile on a specific ArbOS version
```

After you add `sayHi()` to `ArbSys.go` according to the guide in [customize precompile Option 1](./precompiles.mdx#option-1-add-new-methods-to-an-existing-precompile),

> you need continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).

Reviewer: ```suggestion
you need to continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).
```

> you need continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).

Reviewer: ```suggestion
you need to continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).
```

> you need continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).

Reviewer: ```suggestion
you need to continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).
```

> you need continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).

Reviewer: ```suggestion
you need to continue to modify [precompile.go](https://github.com/OffchainLabs/@@nitroRepositorySlug=nitro@@/blob/@@nitroVersionTag=v3.9.9@@/@@nitroPathToPrecompiles=precompiles@@/precompile.go).
```

:::caution

It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./precompiles.mdx#option-5-call-and-modify-state)), you need to initialize it to `0` or `null` value first to avoid a potential re-org on your chain.

:::

> Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS v21.

Reviewer: ```suggestion
Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions that may change the value of `myNumber` before ArbOS v21.
```

> Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS v21.

Reviewer: ```suggestion
Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions that may change the value of `myNumber` before ArbOS v21.
```

> Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS v21.

Reviewer: ```suggestion
Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions that may change the value of `myNumber` before ArbOS v21.
```

> Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS v21.

Reviewer: ```suggestion
Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions that may change the value of `myNumber` before ArbOS v21.
```
